From 7ce0bcb85e22b098a79af4cd7470e3240401f482 Mon Sep 17 00:00:00 2001
From: Steve Lhomme <robux4@ycbcr.xyz>
Date: Fri, 12 Feb 2021 11:10:03 +0100
Subject: [PATCH 07/10] avcodec/mpeg12dec: don't end a slice without
 first_slice

If first_slice is set that means the first slice/field is not started yet. We
should not end the slice. In particular calling hwaccel->end_frame may crash as
we're ending a frame that was not started.

We also need to reset first_slice once the slice_end is finished handling
for this check to work.
---
 libavcodec/mpeg12dec.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/libavcodec/mpeg12dec.c b/libavcodec/mpeg12dec.c
index f22396f66e..39dc481cea 100644
--- a/libavcodec/mpeg12dec.c
+++ b/libavcodec/mpeg12dec.c
@@ -2241,9 +2241,14 @@ static int decode_chunks(AVCodecContext *avctx, AVFrame *picture,
             if (!skip_frame) {
                 mpeg12_execute_slice_threads(avctx, s);
 
-                ret = slice_end(avctx, picture, got_output);
+                if (s->first_slice) // not started yet. don't end it
+                    ret = 0;
+                else
+                    ret = slice_end(avctx, picture, got_output);
                 if (ret < 0)
                     return ret;
+                // slice ended, don't end it again later
+                s->first_slice = 1;
             }
             s2->pict_type = 0;
 
-- 
2.45.1.windows.1

